govcd/vapp_network.go

```diff
@@ -139,3 +143,80 @@ func (vapp *VApp) GetVappNetworkByNameOrId(identifier string, refresh bool) (*ty
 	}
 	return entity.(*types.VAppNetwork), err
 }
+
+// UpdateNetworkNatRules updates vApp networks NAT rules.
+// Returns pointer to types.VAppNetwork or error
+func (vapp *VApp) UpdateNetworkNatRules(networkId string, natRules []*types.NatRule, natType, policy string) (*types.VAppNetwork, error) {
+	task, err := vapp.UpdateNetworkNatRulesAsync(networkId, natRules, natType, policy)
+	if err != nil {
+		return nil, err
+	}
+	err = task.WaitTaskCompletion()
+	if err != nil {
+		return nil, fmt.Errorf("%s", combinedTaskErrorMessage(task.Task, err))
+	}
+
+	return vapp.GetVappNetworkById(networkId, false)
+}
+
+// UpdateNetworkNatRulesAsync asynchronously updates vApp NAT rules.
+// Returns task or error
+func (vapp *VApp) UpdateNetworkNatRulesAsync(networkId string, natRules []*types.NatRule, natType, policy string) (Task, error) {
+	util.Logger.Printf("[TRACE] UpdateNetworkNatRulesAsync with values: id: %s and natRules: %#v", networkId, natRules)
+
+	uuid := extractUuid(networkId)
+	networkToUpdate, err := vapp.GetVappNetworkById(uuid, true)
+	if err != nil {
+		return Task{}, err
+	}
+
+	if networkToUpdate.Configuration.Features == nil {
+		networkToUpdate.Configuration.Features = &types.NetworkFeatures{}
+	}
+	networkToUpdate.Xmlns = types.XMLNamespaceVCloud
+
+	if networkToUpdate.Configuration.Features.NatService == nil {
+		return Task{}, fmt.Errorf("provided network isn't connected to org network or isn't fenced")
+	}
+	networkToUpdate.Configuration.Features.NatService.NatType = natType
+	networkToUpdate.Configuration.Features.NatService.Policy = policy
+	networkToUpdate.Configuration.Features.NatService.NatRule = natRules
+	networkToUpdate.Configuration.Features.NatService.IsEnabled = true
+	networkToUpdate.Configuration.Features.FirewallService.IsEnabled = true
```

Comment:
It seems, that in UI it is possible to edit FW and NAT rules, but leave FW and NAT services disabled. In other words, control enabling/disabling of these services separately. Here we are overwriting the previous state of these services (turning them on). Comments, suggestions?

Similarly, we should check these two fields' behavior in the test below as well.

CC: @dataclouder @Didainius

Reply:
Ty, Linas for noticing this. I remove this as this not needed anymore I added through research phase and haven't removed. FYI NAT can't be enabled if FW isn't enabled. You can see this behaviour in UI.